File: 0912-sort-an-array/heap-sort/python/solution.py
class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:      

        heap_size = 0
        n = len(nums)      

        def max_heapify(idx):       # O(log n)
            nonlocal heap_size
            largest = idx
            left, right = 2 * idx + 1, 2 * idx + 2
            
            if left < heap_size and nums[left] > nums[largest]:
                largest = left   
            if right < heap_size and nums[right] > nums[largest]:
                largest = right
            
            if largest != idx:
                nums[largest], nums[idx] = nums[idx], nums[largest]   # swap
                max_heapify(largest)

        def build_heap():           # O(n)
            nonlocal heap_size
            heap_size = n
            for idx in range(len(nums) // 2 - 1, -1 , -1):
                max_heapify(idx)
        
        def heap_sort():            # O(n log n)
            nonlocal heap_size
            build_heap()                        
            for idx in range(n - 1, 0, -1):   # from n-1 down to 1 (leaving index 0, which ends up in correct place)
                nums[0], nums[idx] = nums[idx], nums[0]
                heap_size -= 1
                max_heapify(0)         

        heap_sort() 
        return nums

File: 0912-sort-an-array/heap-sort/cpp/solution.cpp
class Solution {
private:
    void maxHeapify(vector<int>& nums, int idx, int heap_size) {   // O(log n)
        int largest = idx;
        int left = 2 * idx + 1;
        int right = 2 * idx + 2;

        if (left < heap_size && nums[left] > nums[largest]) {
            largest = left;
        }
        if (right < heap_size && nums[right] > nums[largest]) {
            largest = right;
        }

        if (largest != idx) {
            swap(nums[idx], nums[largest]);
            maxHeapify(nums, largest, heap_size);
        }
    }

    void buildHeap(vector<int>& nums, int heap_size) {    // O(n)
        for (int idx = heap_size / 2 - 1; idx >= 0; idx--) {
            maxHeapify(nums, idx, heap_size);
        }
    }

    void heapSort(vector<int>& nums) { // O(n log n)
        int heap_size = nums.size();
        buildHeap(nums, heap_size);
        
        for (int idx = nums.size() - 1; idx > 0; idx--) {
            swap(nums[0], nums[idx]);
            heap_size--;
            maxHeapify(nums, 0, heap_size);
        }
    }

public:
    vector<int> sortArray(vector<int>& nums) {
        heapSort(nums);
        return nums;
    }
};
